pkgs/development/libraries/flutter-engine/default.nix

```diff
@@ -0,0 +1,283 @@
+            tar -xf $toolchain -C $out/src/build/linux/debian_sid_${toolchainArch}-sysroot
+            cd $out
+            mkdir -p $out/src/third_party/dart/.dart_tool
+            cp -r --no-preserve=ownership,mode $dartPackageConfig $out/src/third_party/dart/.dart_tool/package_config.json
+            python3 $out/src/third_party/dart/tools/generate_package_config.py
+            python3 $out/src/third_party/dart/tools/generate_sdk_version_file.py
+            python3 $out/src/tools/remove_stale_pyc_files.py src/tools
+            python3 $out/src/flutter/tools/pub_get_offline.py
+          '';
+        };
+      in
+      stdenvNoCC.mkDerivation rec {
+        pname = "flutter-engine-${runtimeMode}";
+        version = builtins.substring 0 7 args.version;
+        inherit enableParallelBuilding runtimeMode src;
+
+        nativeBuildInputs = [ git python3 pkg-config ninja stdenv.cc.libc ];
+
+        requiredSystemFeatures = optional enableParallelBuilding "big-parallel";
+
+        ## FIXME: create a sysroot toolchain for Nix and use it instead
+        ## TODO: generate for other platforms based on "targetPlatform"
+        configurePhase = ''
+          ./src/flutter/tools/gn \
+            --runtime-mode $runtimeMode \
+            --no-build-glfw-shell \
+            --prebuilt-dart-sdk \
+            --embedder-for-target \
+            --no-goma \
+            --out-dir $out/lib/flutter \
+            --target-dir $runtimeMode
+        '';
+
+        ## TODO: add more builds and possibly better cross-compiling
+        buildPhase = ''
+          cd $out/lib/flutter/out/$runtimeMode
+          export TERM=dumb
+          # Comes from "pkgs/development/tools/build-managers/ninja/setup-hook.sh"
+          local buildCores=1
+          if [ "''${enableParallelBuilding-1}" ]; then
+            buildCores="$NIX_BUILD_CORES"
+          fi
+          local flagsArray=(
+            -j$buildCores
+            $ninjaFlags "''${ninjaFlagsArray[@]}"
+          )
+          echo "Building flatc"
+          ninja flatc "''${flagsArray[@]}"
+          patchelf --set-interpreter ${stdenv.cc.libc}/lib/${interpreter} flatc
+          echo "Building blobcat"
+          ninja blobcat "''${flagsArray[@]}"
+          patchelf --set-interpreter ${stdenv.cc.libc}/lib/${interpreter} blobcat
+          echo "Building gen_snapshot"
+          ninja gen_snapshot "''${flagsArray[@]}"
+          patchelf --set-interpreter ${stdenv.cc.libc}/lib/${interpreter} gen_snapshot
+          echo "Building impellerc"
+          ninja impellerc "''${flagsArray[@]}"
+          patchelf --set-interpreter ${stdenv.cc.libc}/lib/${interpreter} impellerc
+          echo "Building Flutter Engine library"
+          ninja flutter_engine_library "''${flagsArray[@]}"
+          echo "Building icudtl.dat"
+          ninja icudtl.dat "''${flagsArray[@]}"
+          ninja flutter_embedder.h "''${flagsArray[@]}"
```

Comment:
refactor

```sh
          for target in flatc blobcat gen_snapshot etc etc; do
            echo "Building $target"
            ninja $target "''${flagsArray[@]}"
          done
```

Reply:
I've got 2 of these for loops since we gotta patch the binaries in the first half.